Increase TLSRoute hostnames limit from 16 to 1024

alexanderstephan opened the pull request:
<!--  Thanks for sending a pull request! Here are some tips for you:

1. If this is your first time contributing to Gateway API, please read our
   developer guide (https://gateway-api.sigs.k8s.io/contributing/devguide/)
   and our community page (https://gateway-api.sigs.k8s.io/contributing/).
2. If this is your first time contributing to a Kubernetes project, please read
   our contributor guidelines:
   https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution
3. Please label this pull request according to what type of issue you are
   addressing, especially if this is a release targeted pull request. For
   reference on required PR/issue labels, read here:
   https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
4. If you want *faster* PR reviews, read how:
   https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it:
   https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
6. If this PR includes a new GEP please make sure you've followed the process
   outlined in our GEP overview, as this will help the community to ensure the
   best chance of positive outcomes for your proposal:
   https://gateway-api.sigs.k8s.io/geps/overview/#process
-->

**What type of PR is this?**
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind gep
/kind test

Optionally add one or more of the following kinds if applicable:
/kind deprecation
/kind failing-test
/kind flake
/kind regression
/area conformance-test
/area conformance-machinery
-->
/kind feature
/kind documentation

**What this PR does / why we need it**:
This PR updates the TLSRoute CRD validation rationale and proposes increasing the `maxItem` bound for hostnames from 16 to 4096. This change is proposed only to accommodate very large orgs. In large orgs, the current limit of 16, this can lead to hundred of thousands TLSRoute objects. These objects multiply storage, watch traffic, and controller memory/CPU, which drives up API-server latency and risks OOMs and instability.

In a similar PR the limit has been increased: https://github.com/kubernetes-sigs/gateway-api/pull/3205/
To safely deploy the change the author employed a XValidation rule. For my case, such a rule would likely be rejected for being too complex. One idea would be to add the validation to a Validating Webhook. However, the webhook as been removed:

```
The validating webhook has been removed. CEL validation is now built-in to CRDs and replaces the webhook. (#2595, @robscott)
```
Do you have any ideas on how to solve this issue? I would be happy for further assistance on how to tackle this. 

Rationale:
- Hostname size on the wire: DNS RFCs limit a full domain name to 255 octets on the wire (see RFC 2181 §11). In practice this means up to roughly 253 printable characters in common textual form (no trailing dot); for SNI and Gateway API usage the ACE/Punycode representation should be used for IDNs and the same limits apply.
  - RFC reference: https://datatracker.ietf.org/doc/html/rfc2181#section-11
  - TLS/SNI reference: https://datatracker.ietf.org/doc/html/rfc6066#section-3
- etcd / API server request-size guidance: etcd dev-guide (v3.3) documents a practical per-request/message size ~1.5 MiB. kube-apiserver also commonly enforces a request-size cap). See: https://etcd.io/docs/v3.3/dev-guide/limit/#:~:text=etcd%20is%20designed%20to%20handle,any%20request%20is%201.5%20MiB
- Based on these assumptions, I did the following napkin calculation: I assume a base overhead ≈ 4 KB, then I get  (1,572,864 − 4,096) / 256 = 1,568,768 / 256 = 6,128, so 4096 could be a conservative upper limit. 

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
No issue yet. 

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, please enter a release note below:
-->

```release-note
The TLSRoute CRD validation has been adjusted to allow up to 4096 hostnames and rules per TLSRoute resource. Operators must validate kube-apiserver, etcd and Gateway controller behavior with representative manifests prior to enabling the new limit in production.
```

youngnick commented:
Hi @alexanderstephan, I'd prefer to keep the increase to a smaller one. 16 to 4096 is 8 orders of magnitude, that seems like too big a jump.

To rerun some of your calculations:

- maximum hostname size is 256 characters
- we need to allow at least 6, maybe 8 characters of overhead per hostname (indenting, `- ` and maybe quotes). Let's go worst-case and call it 8.


4096 is, as you said : (256+8)*4096 = 1,081,344 bytes. That doesn't leave much overhead.

If we increase by 6 orders of magnitude, and go with 1024, then we have: (256+8)*1024 = 270,336 bytes.

I think that's a better place to start, and gives you a _significant_ increase, while keeping plenty of space for further changes.

I _do_ think you may find other limitations inside implementations when using large numbers of `hostname` entries like this, though. 

@rikatz, ping for your thoughts here.

youngnick commented:
/remove-lifecycle stale

alexanderstephan commented:
Hi @youngnick, Thanks for looking into it again! Even 1024 would improve the situation a lot! Then, I would change it and also try to investigate for further side effects of this change.

rikatz commented:
Can we make 128 first, and then move forward? I agree with Nick, making it 4096 is too much, and I think even 1024 is a lot and doesn´t give us much room to mistakes.

Thanks!

Garfield96 commented:
Multitenant environments can have 10000+ tenants, which would result in a lot of TLSRoute resources if only 128 entries are allowed. In my opinion, resource size is not big concern. If a user decides to have a lot of hostnames, it is the users responsibility to make sure it doesn't exceed the ETCD limits. In addition, the ETCD limit of 1.5 MiB is just the default. Users may decide to increase or decrease it, which would invalidate the calculations. Varying lengths of hostnames make the size impact even more unpredictable. I agree with @youngnick that other infrastructure components, like certificate handlers, might not be able to handle a very large amount of hostnames. However, this problem is completely unpredictable and I believe this also falls into the responsibility of the developer.

Hence, I support the proposal of @youngnick to start with 1024 entries.

rikatz commented:
sounds good, I think we can at least stick with it on 1024 then

Can you please rebase and fix the conflicts?

/ok-to-test

alexanderstephan committed:
Increase TLSRoute hostnames limit to 1024

alexanderstephan commented:
Hi @rikatz, I updated all occurrences I found and rebased onto the `main` branch. Please verify whether all changes are at right place as I lack the familiarity with API structure. Thanks!

youngnick commented:
Note that, even once it's merged, this will not take effect until the release of v1.6, due end of June. We don't backport changes to earlier versions.

rikatz commented:
/lgtm
/approve
thanks. As Nick pointed, this will be part of 1.6 release

alexanderstephan commented:
@youngnick, @rikatz 
No issue, we're not in a hurry. But very happy to see this coming, thanks guys! 🎉
